### PR TITLE
executor, session: refine insert unsigned bigint autoIncreID

### PR DESCRIPTION
**Pull request by XuHuaiyu**
Refer: https://github.com/pingcap/parser/pull/34

### What problem does this PR solve? 
Before this commit, the behavior of the following sqls in TiDB is wrong.

- unsigned auto_increment integer column:
``` sql
-- We cannot insert an unsigned integer value which is larger than MaxInt64 into a signed auto_increment column.
tidb> create table unsigned_autoid(`auto_inc_id` bigint unsigned not null unique key auto_increment);
tidb> insert into unsigned_autoid values(9223372036854775808);
ERROR 1690 (22003): constant 9223372036854775808 overflows bigint
```

- signed auto_increment integer column:
``` sql
-- The value inserted is wrong if the auto_increment col is MaxInt64 already.
tidb> create table signed_auto_inc(auto_inc_col bigint signed auto_increment, key(au
to_inc_col));
Query OK, 0 rows affected (0.01 sec)
tidb> insert into signed_auto_inc value(cast((1<<63)-1 as signed));
Query OK, 1 row affected (0.00 sec)
tidb> insert into signed_auto_inc values();
Query OK, 1 row affected (0.00 sec)
tidb> insert into signed_auto_inc values();
Query OK, 1 row affected (0.00 sec)
tidb> select * from signed_auto_inc;
+----------------------+
| auto_inc_col         |
+----------------------+
|  9223372036854775807 |
| -9223372036854775807 |
| -9223372036854775805 |
+----------------------+
3 rows in set (0.00 sec)
```

### What is changed and how it works?

1. Add check for the unsigned flag for allocator.
2. Add check for MaxUint64 and MaxInt64.

Note: After this commit, we can **promise the correctness** for the behavior when inserting a specific value or an auto-generated value into the auto-increment column.
**BUT** it is not 100% consistent with MySQL now, we :

- unsigned auto_increment id
``` sql
create table unsigned_autoid(`auto_inc_id` bigint unsigned not null unique key auto_increment);

-- MySQL:
mysql> insert into unsigned_autoid values(18446744073709551614);
Query OK, 1 row affected (0.00 sec)
mysql> insert into unsigned_autoid values();
ERROR 1467 (HY000): Failed to read auto-increment value from storage engine
mysql> insert into unsigned_autoid values(18446744073709551615); # MySQL performs well.
Query OK, 1 row affected (0.00 sec)

--TiDB:
tidb> insert into unsigned_autoid values(18446744073709551614);
Query OK, 1 row affected (0.00 sec)
tidb> insert into unsigned_autoid values();
ERROR 1467 (HY000): Failed to read auto-increment value from storage engine
tidb> insert into unsigned_autoid values(18446744073709551615);
ERROR 1467 (HY000): Failed to read auto-increment value from storage engine
```

- signed auto_increment id
``` sql
create table signed_autoid(`auto_inc_id` bigint signed not null unique key auto_increment);

-- MySQL:
mysql> insert into signed_autoid values(9223372036854775806);
Query OK, 1 row affected (0.00 sec)
mysql> insert into signed_autoid values(); # MySQL performs well.
Query OK, 1 row affected (0.00 sec)
mysql> insert into signed_autoid values(); # MySQL returns DupEntry error
ERROR 1062 (23000): Duplicate entry '9223372036854775807' for key 'auto_inc_id'

--TiDB:
tidb> insert into signed_autoid values(9223372036854775806);
Query OK, 1 row affected (0.00 sec)
tidb> insert into signed_autoid values();
ERROR 1467 (HY000): Failed to read auto-increment value from storage engine
tidb> insert into signed_autoid values();
ERROR 1467 (HY000): Failed to read auto-increment value from storage engine
```

### Check List 

Tests 

 - Integration test

Code changes

 - Has exported function/method change

Side effects

none

Related changes

 - Need to cherry-pick to the release branch

release-2.0 release-2.1

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8181)

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
@zimulala @jackysp PTAL

File: session/session_test.go [hunk 1]
```diff
@@ -704,6 +704,11 @@ func (s *testSessionSuite) TestAutoIncrementID(c *C) {
 
 	tk.MustQuery("select last_insert_id(20)").Check(testkit.Rows(fmt.Sprint(20)))
 	tk.MustQuery("select last_insert_id()").Check(testkit.Rows(fmt.Sprint(20)))
+
+	tk.MustExec("drop table if exists autoid")
+	tk.MustExec("create table autoid(`auto_inc_id` bigint(20) UNSIGNED NOT NULL AUTO_INCREMENT,UNIQUE KEY `auto_inc_id` (`auto_inc_id`))")
+	tk.MustExec("insert into autoid values(9223372036854775808);")
+	tk.MustQuery("select * from autoid").Check(testkit.Rows("9223372036854775808"))
```
Comment (jackysp) [inline session/session_test.go]: How about insert auto values after this? e.g. `insert into autid values ();`

Comment (jackysp) [reply session/session_test.go]:
```sql
mysql> create table autoid(`auto_inc_id` bigint(20) UNSIGNED NOT NULL AUTO_INCREMENT,UNIQUE KEY `auto_inc_id` (`auto_inc_id`));
Query OK, 0 rows affected (0.02 sec)

mysql> insert into autoid values(9223372036854775808);
Query OK, 1 row affected (0.00 sec)

mysql> select * from autoid;
+---------------------+
| auto_inc_id         |
+---------------------+
| 9223372036854775808 |
+---------------------+
1 row in set (0.00 sec)

mysql> insert into autoid values();
Query OK, 1 row affected (0.00 sec)

mysql> select * from autoid;
+---------------------+
| auto_inc_id         |
+---------------------+
| 9223372036854775808 |
|                   2 |
+---------------------+
2 rows in set (0.00 sec)
```

File: executor/insert_common.go [hunk 2]
```diff
@@ -446,9 +446,19 @@ func (e *InsertValues) adjustAutoIncrementDatum(d types.Datum, hasValue bool, c
...
+			uintD, err = d.ConvertTo(sc, &c.FieldType)
+			if e.filterErr(err) != nil {
+				return types.Datum{}, err
+			}
+			recordID = int64(uintD.GetUint64())
+		} else {
+			recordID, err = d.ToInt64(sc)
+			if e.filterErr(err) != nil {
+				return types.Datum{}, errors.Trace(err)
+			}
```
Comment (eurekaka) [inline executor/insert_common.go]: `ConvertTo` checks `mysql.HasUnsignedFlag`, how about:
```go
datum, err := d.ConvertTo(sc, &c.FieldType)
if e.filterErr(err) != nil {
    return types.Datum{}, err
}
recordID = datum.GetInt64()
```

**Comment by XuHuaiyu** [comment]
PTAL @jackysp @zimulala @winkyao

**Comment by tiancaiamao** [comment]
https://github.com/pingcap/parser/pull/34 is merged.
Please update the `go.mod` file to use the latest parser

File: session/session_test.go [hunk 3]
```diff
@@ -704,6 +705,29 @@ func (s *testSessionSuite) TestAutoIncrementID(c *C) {
...
+	c.Assert(terror.ErrorEqual(err, autoid.ErrAutoincReadFailed), IsTrue)
+	// FixMe: MySQL works fine with the this sql.
+	_, err = tk.Exec("insert into autoid values(18446744073709551615)")
+	c.Assert(terror.ErrorEqual(err, autoid.ErrAutoincReadFailed), IsTrue)
+
+	// Corner cases for signed bigint auto_increment Columns.
+	tk.MustExec("drop table if exists autoid")
+	tk.MustExec("create table autoid(`auto_inc_id` bigint(20) NOT NULL AUTO_INCREMENT,UNIQUE KEY `auto_inc_id` (`auto_inc_id`))")
+	tk.MustExec("insert into autoid values(9223372036854775806);")
+	tk.MustExec("insert into autoid values();")
```
Comment (crazycs520) [inline session/session_test.go]: the test has a problem:
```sql
create table autoid(`auto_inc_id` bigint(20) NOT NULL AUTO_INCREMENT,UNIQUE KEY `auto_inc_id` (`auto_inc_id`));
insert into autoid values(9223372036854775806);
insert into autoid values();
mysql root@127.0.0.1:test> select * from autoid
+---------------------+
| auto_inc_id         |
+---------------------+
| 9223372036854775807 |
+---------------------+
1 row in set
Time: 0.006s
mysql root@127.0.0.1:test> admin check table autoid;
(8003, u'autoid err:[admin:1]index:&admin.RecordData{Handle:9223372036854775807, Values:[]types.Datum{types.Datum{k:0x1, collation:0x0, decimal:0x0, length:0x0, i:9223372036854775806, b:[]uint8(nil), x:interface {}(nil)}}} != record:&admin.RecordData{Handle:9223372036854775807, Values:[]types.Datum{types.Datum{k:0x1, collation:0x0, decimal:0x0, length:0x0, i:9223372036854775807, b:[]uint8(nil), x:interface {}(nil)}}}')
```

File: meta/autoid/autoid.go [hunk 4]
```diff
@@ -165,21 +200,38 @@ func (alloc *allocator) Alloc(tableID int64) (int64, error) {
...
+			return 0, err
 		}
 		alloc.base, alloc.end = newBase, newEnd
 	}
+	log.Warning(uint64(alloc.base), uint64(alloc.end))
 
-	alloc.base++
-	log.Debugf("[kv] Alloc id %d, table ID:%d, from %p, database ID:%d", alloc.base, tableID, alloc, alloc.dbID)
+	if alloc.isUnsigned {
+		if uint64(alloc.base) == math.MaxUint64 {
```
Comment (crazycs520) [inline meta/autoid/autoid.go]: move out of if. !alloc.isUnsigned need this check too.

**Comment by crazycs520** [comment]
I have a question: Why `auto_increment` and `_tidb_rowid` both use the same `allocator`?
@zimulala

File: meta/autoid/autoid.go [hunk 5]
```diff
@@ -137,13 +137,12 @@ func (alloc *allocator) Rebase(tableID, requiredBase int64, allocIDs bool) error
 		if allocIDs {
 			if alloc.isUnsigned {
```
Comment (tiancaiamao) [inline meta/autoid/autoid.go]: You want to prevent overflow? 
Should we 
```
if uNewBase > (math.Uint64 - step) {
   panic
}
```

Comment (XuHuaiyu) [reply meta/autoid/autoid.go]: Yes,
but why we panic here? @tiancaiamao

Comment (tiancaiamao) [reply meta/autoid/autoid.go]: Well, maybe error or warning...
I mean, if the ID overflow, it should not be changed to math.Uint64

Comment (tiancaiamao) [reply meta/autoid/autoid.go]: If alloc.end overflow, what will happen?

Comment (XuHuaiyu) [reply meta/autoid/autoid.go]: @tiancaiamao 
The result of TiDB master:
``` sql
tidb> create table t(a bigint signed not null auto_increment, unique key(a));
Query OK, 0 rows affected (0.02 sec)

tidb> insert into t value(9223372036854775806);
Query OK, 1 row affected (0.00 sec)

tidb> insert into t value();
Query OK, 1 row affected (0.00 sec)

tidb> insert into t value();
Query OK, 1 row affected (0.00 sec)

tidb> insert into t value();
Query OK, 1 row affected (0.00 sec)

tidb> insert into t value();
Query OK, 1 row affected (0.00 sec)

tidb> select * from t;
+----------------------+
| a                    |
+----------------------+
| -9223372036854775808 |
| -9223372036854775806 |
| -9223372036854775804 |
| -9223372036854775802 |
|  9223372036854775806 |
+----------------------+
5 rows in set (0.00 sec)
```

Comment (tiancaiamao) [reply meta/autoid/autoid.go]: Overwrite the old row id is dangerous for data consistency.
Is it acceptable to change the behavior to throw error?

Comment (XuHuaiyu) [reply meta/autoid/autoid.go]: As a tradeoff, https://github.com/pingcap/tidb/pull/8181#pullrequestreview-174023063 may be better?

File: meta/autoid/autoid.go [hunk 6]
```diff
@@ -165,21 +200,38 @@ func (alloc *allocator) Alloc(tableID int64) (int64, error) {
...
 			return nil
 		})
 		metrics.AutoIDHistogram.WithLabelValues(metrics.TableAutoIDAlloc, metrics.RetLabel(err)).Observe(time.Since(startTime).Seconds())
 		if err != nil {
-			return 0, errors.Trace(err)
+			return 0, err
 		}
 		alloc.base, alloc.end = newBase, newEnd
 	}
+	log.Warning(uint64(alloc.base), uint64(alloc.end))
```
Comment (tiancaiamao) [inline meta/autoid/autoid.go]: log.Warning what ?

Comment (XuHuaiyu) [reply meta/autoid/autoid.go]: Just for debugging, can be removed.

File: meta/autoid/autoid.go (hunk 5 above)
Comment (tiancaiamao) [reply meta/autoid/autoid.go]: I think throw error is simple, easier, and more robust.

File: meta/autoid/autoid.go (hunk 6 above)
Comment (XuHuaiyu) [reply meta/autoid/autoid.go]: Please check https://github.com/pingcap/tidb/pull/8181/commits/758fd314f1011612482f51e68c19b923c99126eb.

**Comment by XuHuaiyu** [comment]
comments addressed
PTAL @zimulala @tiancaiamao @zz-jason @crazycs520 @jackysp @eurekaka

File: meta/autoid/autoid.go [hunk 7]
```diff
@@ -91,25 +92,65 @@ func (alloc *allocator) NextGlobalAutoID(tableID int64) (int64, error) {
...
+		currentEnd, err1 := m.GetAutoTableID(alloc.dbID, tableID)
+		if err1 != nil {
+			return errors.Trace(err1)
+		}
+		uCurrentEnd := uint64(currentEnd)
+		if allocIDs {
+			newBase = mathutil.MaxUint64(uCurrentEnd, requiredBase)
+			newEnd = mathutil.MinUint64(math.MaxUint64-uint64(step), newBase) + uint64(step)
+		} else {
+			if !alloc.isUnsigned && uCurrentEnd >= requiredBase {
```
Comment (crazycs520) [inline meta/autoid/autoid.go]: `if !alloc.isUnsigned` check is redundant, already check in line#203

File: meta/autoid/autoid.go [hunk 8]
```diff
@@ -123,9 +164,9 @@ func (alloc *allocator) Rebase(tableID, requiredBase int64, allocIDs bool) error
 		}
 		if allocIDs {
 			newBase = mathutil.MaxInt64(currentEnd, requiredBase)
-			newEnd = newBase + step
+			newEnd = mathutil.MinInt64(math.MaxInt64-step, newBase) + step
 		} else {
-			if currentEnd >= requiredBase {
+			if !alloc.isUnsigned && currentEnd >= requiredBase {
```
Comment (crazycs520) [inline meta/autoid/autoid.go]: ditto.

**Comment by XuHuaiyu** [comment]
@crazycs520 Ok, test will fail now, I'll update go.mod and run it again.

File: meta/autoid/autoid.go [hunk 9]
```diff
@@ -123,9 +164,9 @@ func (alloc *allocator) Rebase(tableID, requiredBase int64, allocIDs bool) error
 		}
 		if allocIDs {
 			newBase = mathutil.MaxInt64(currentEnd, requiredBase)
-			newEnd = newBase + step
+			newEnd = mathutil.MinInt64(math.MaxInt64-step, newBase) + step
```
Comment (winkyao) [inline meta/autoid/autoid.go]: Why not math.MaxUint64?

File: meta/autoid/autoid.go (hunk 8 above)
Comment (winkyao) [reply meta/autoid/autoid.go]: !alloc.isUnsigned will always be true here?

File: session/session_test.go [hunk 10]
```diff
@@ -755,6 +756,30 @@ func (s *testSessionSuite) TestAutoIncrementID(c *C) {
...
+
+	// Corner cases for signed bigint auto_increment Columns.
+	tk.MustExec("drop table if exists autoid")
+	tk.MustExec("create table autoid(`auto_inc_id` bigint(20) NOT NULL AUTO_INCREMENT,UNIQUE KEY `auto_inc_id` (`auto_inc_id`))")
+	tk.MustExec("insert into autoid values(9223372036854775806);")
+	tk.MustQuery("select auto_inc_id, _tidb_rowid from autoid").Check(testkit.Rows("9223372036854775806 9223372036854775807"))
+	_, err = tk.Exec("insert into autoid values();")
+	c.Assert(terror.ErrorEqual(err, autoid.ErrAutoincReadFailed), IsTrue)
+	tk.MustQuery("select auto_inc_id, _tidb_rowid from autoid").Check(testkit.Rows("9223372036854775806 9223372036854775807"))
+	tk.MustQuery("select auto_inc_id, _tidb_rowid from autoid use index(auto_inc_id)").Check(testkit.Rows("9223372036854775806 9223372036854775807"))
```
Comment (winkyao) [inline session/session_test.go]: Add more tests case here, like allocate more than `step` ids, then we can know your changes work when it occurs rebase.

File: meta/autoid/autoid.go (hunk 7 above)
Comment (winkyao) [reply meta/autoid/autoid.go]: I think check !alloc.isUnsigned is not necessary.

**Comment by XuHuaiyu** [comment]
PTAL @winkyao @crazycs520

File: ddl/ddl_api.go [hunk 11]
```diff
@@ -1021,7 +1021,11 @@ func checkCharsetAndCollation(cs string, co string) error {
 // handleAutoIncID handles auto_increment option in DDL. It creates a ID counter for the table and initiates the counter to a proper value.
 // For example if the option sets auto_increment to 10. The counter will be set to 9. So the next allocated ID will be 10.
 func (d *ddl) handleAutoIncID(tbInfo *model.TableInfo, schemaID int64) error {
-	alloc := autoid.NewAllocator(d.store, tbInfo.GetDBID(schemaID))
+	isAutoIncColUnsigned := false
```
Comment (jackysp) [inline ddl/ddl_api.go]: I think it is better to extract these lines to a function.

Comment (XuHuaiyu) [reply ddl/ddl_api.go]: you mean line1024~1028?

Comment (jackysp) [reply ddl/ddl_api.go]: Yes, I see some similar code in this PR.

**Comment by XuHuaiyu** [comment]
PTAL @winkyao @jackysp @crazycs520

**Comment by XuHuaiyu** [comment]
should be merged with https://github.com/pingcap/parser/pull/79

**Comment by winkyao** [comment]
@XuHuaiyu CI failed.

**Pull request merged**